thorlcr/thorutil/thormisc.cpp

```diff
@@ -1670,3 +1670,31 @@ StringBuffer &getExpertOptString(const char *opt, StringBuffer &out)
     globals->getProp(xpath, out);
     return out;
 }
+
+// NB: in practice this is for containerized only
+void applyResourcedCPUs(const char *resourceSection)
+{
+    unsigned maxActivityCores = getExpertOptInt64(THOROPT_MAX_ACTIVITY_CORES);
+    if (maxActivityCores)
+        return;
+    VStringBuffer xpath("%s/@cpu", resourceSection);
+    if (!globals->hasProp(xpath))
+        return;
+    const char *cpusText = globals->queryProp(xpath);
+    double cpus = friendlyCPUToDecimal(cpusText);
+    if (0.0 == cpus)
+        throw makeStringExceptionV(0, "Invalid number of %s.cpu: %s", resourceSection, cpusText);
+    unsigned __int64 cpusI = (unsigned __int64)cpus;
+    if (cpus != (double)cpusI)
+    {
+        OWARNLOG("Fractional number of CPUs '%s' specified in %s.cpu. May cause poor performance", cpusText, resourceSection);
+        // round up
+        cpusI++;
+    }
+
+    StringBuffer optPath;
+    getExpertOptPath(nullptr, optPath);
+    ensurePTree(globals, optPath);
+    getExpertOptPath(THOROPT_MAX_ACTIVITY_CORES, optPath.clear());
```

Comment:
I expected this to override the result of getAffinityCpus().  Otherwise things like sort, the task manager roxie strands, workflow will still be unlimited - potentially causing significant throttling.

Reply:
I think everywhere in Thor should pick up and use this setting, if there are places that don't then I think that's a hole in the design because, irrespective of this change, the existing impl. should allow global,workunit and per activity (via HINT) to set the maxCores for any activity. Each activity impl. should be picking up this property.

sort etc. does pick it up.

But you're right if there are lower level classes outside thor that use getAffinityCpus() directly, they will continue to use too many threads.

I'll look at setting it so that getAffinityCpus() returns the value instead.